File: client/src/modules/logState/LogState.js
var LogState = cc.Class.extend({

    ctor: function () {
        this.flag = false;
        this.dirState = "../logClient/logState/user" + gv.user.id + "/"
        this.dirSumState = "../logClient/logSumState/user" + gv.user.id + "/"
        jsb.fileUtils.createDirectory(this.dirState)
        jsb.fileUtils.createDirectory(this.dirSumState)

        this.deltaMonster = 0.1;
        this.deltaTree = 0.1;
        this.deltaTower = 0.1;
        this.deltaBullet = 0.1;
        this.deltaEnergy = 0.1;
    },

    logToFile: function (gameMgr, countLoop) {
        var content = ""

        // Log monster
        content += "Number of monster: " + gameMgr.listLogicMonster.length
        content += "\n   [id - x - y - current speed - current HP]"
        var sumMonster = 0
        let listMonster = gameMgr.getListMonsterInCellNeighbor(2, 5, 10);
        for (var i = 0; i < listMonster.length; i++) {
            content += "\n" + i + ". ["
                + listMonster[i].id + " - "
                + listMonster[i].position.x.toFixed(3) + " - "
                + listMonster[i].position.y.toFixed(3) + " - "
                + listMonster[i].curSpeed.toFixed(3) + " - "
                + listMonster[i].currentHP.toFixed(3) + "]"

            sumMonster += listMonster[i].id
                + listMonster[i].position.x
                + listMonster[i].position.y
                + listMonster[i].curSpeed
                + listMonster[i].currentHP
        }

        // Log tree
        content += "\n\n\nNumber of tree: " + gameMgr.listLogicTreeObstacle.length
        content += "\n   [id - x - y - current HP]"
        var sumTree = 0
        let listTree = gameMgr.listLogicTreeObstacle;
        for (var i = 0; i < listTree.length; i++) {
            content += "\n" + i + ". ["
                + listTree[i].id + " - "
                + listTree[i].position.x.toFixed(3) + " - "
                + listTree[i].position.y.toFixed(3) + " - "
                + listTree[i].currentHP.toFixed(3) + "]"

            sumTree += listTree[i].id
                + listTree[i].position.x
                + listTree[i].position.y
                + listTree[i].currentHP
        }

        // Log tower
        content += "\n\n\nNumber of tower: " + gameMgr.listLogicTower.length
        content += "\n   [towerID - x - y - evolution]"
        var sumTower = 0
        for (var i = 0; i < gameMgr.listLogicTower.length; i++) {
            content += "\n" + i + ". ["
                + gameMgr.listLogicTower[i].towerID + " - "
                + gameMgr.listLogicTower[i].x.toFixed(3) + " - "
                + gameMgr.listLogicTower[i].y.toFixed(3) + " - "
                + gameMgr.listLogicTower[i].evolution + "]"

            sumTower += gameMgr.listLogicTower[i].towerID
                + gameMgr.listLogicTower[i].x
                + gameMgr.listLogicTower[i].y
                + gameMgr.listLogicTower[i].evolution

            let tower = gameMgr.listLogicTower[i];
            if (tower instanceof TowerLogicAttack) {
                let targetID = tower.getTarget() instanceof LogicMonster ? tower.getTarget().getId() : -1;
                content += " - target: " + targetID;
                sumTower += targetID;
                content += " - range: " + tower.getRange().toFixed(3);
                sumTower += tower.getRange();
                content += " - damage: " + tower.getDamage().toFixed(3);
                sumTower += tower.getDamage();
                content += " - attackSpeed: " + tower.getAttackSpeed().toFixed(3);
                sumTower += tower.getAttackSpeed();
            }
        }

        // Log bullet
        content += "\n\n\nNumber of bullet: " + gameMgr.listLogicBullet.length
        content += "\n   [x - y - damage - speed]"
        var sumBullet = 0
        for (var i = 0; i < gameMgr.listLogicBullet.length; i++) {
            content += "\n" + i + ". ["
                + gameMgr.listLogicBullet[i].x.toFixed(3) + " - "
                + gameMgr.listLogicBullet[i].y.toFixed(3) + " - "
                + gameMgr.listLogicBullet[i].damage.toFixed(3) + " - "
                + gameMgr.listLogicBullet[i].speed.toFixed(3) + "]"

            sumBullet += gameMgr.listLogicBullet[i].x
                + gameMgr.listLogicBullet[i].y
                + gameMgr.listLogicBullet[i].damage
                + gameMgr.listLogicBullet[i].speed
        }

        var energy = gameMgr.currentEnergy;

        // Get path
        var pathState = this.dirState + "/"+ countLoop +"-client-stateFrame.txt"
        var pathSumState = this.dirSumState + "/"+ countLoop +"-client-sumStateFrame.txt"

        this.dirSumStateServer = "../logServer/logSumState/user" + gv.user.id + "/"
        var pathSumStateServer = this.dirSumStateServer + "/"+ countLoop +"-server-sumStateFrame.txt"

        // Compare with server state
        var sumStateServer = jsb.fileUtils.getStringFromFile(pathSumStateServer)
        var sum = sumStateServer.split("\n");
        var svSumMonster = Number(sum[0])
        var svSumTree = Number(sum[1])
        var svSumTower = Number(sum[2])
        var svSumBullet = Number(sum[3])
        var svEnergy = Number(sum[4])

        this.logError(sumMonster, svSumMonster, countLoop, "Monster", this.deltaMonster)
        this.logError(sumTree, svSumTree, countLoop, "Tree", this.deltaTree)
        this.logError(sumTower, svSumTower, countLoop, "Tower", this.deltaTower)
        this.logError(sumBullet, svSumBullet, countLoop, "Bullet", this.deltaBullet)
        this.logError(energy, svEnergy, countLoop, "Energy", this.deltaEnergy)

        // Round
        sumMonster = Utils.round(sumMonster).toFixed(3);
        sumTree = Utils.round(sumTree).toFixed(3);
        sumTower = Utils.round(sumTower).toFixed(3);
        sumBullet = Utils.round(sumBullet).toFixed(3);
        energy = energy;
        var contentSum = sumMonster + "\n" + sumTree + "\n" +  sumTower + "\n" + sumBullet + "\n" + energy;

        // Log state to file
        jsb.fileUtils.writeStringToFile(content, pathState);
        jsb.fileUtils.writeStringToFile(contentSum, pathSumState);
    },

    logError: function (client, server, countLoop, txt, delta) {
        if (Math.abs(client - server) > delta) {
            cc.log("\n" + txt + " - USER " + gv.user.id + " - ERROR STATE AT FRAME: " + countLoop + ", " + Math.abs(client - server) + "\n")

            switch (txt) {
                case "Monster": {
                    this.deltaMonster = delta + Math.abs(client - server)
                    break
                }
                case "Tree": {
                    this.deltaTree = delta + Math.abs(client - server)
                    break
                }
                case "Tower": {
                    this.deltaTower = delta + Math.abs(client - server)
                    break
                }
                case "Bullet": {
                    this.deltaBullet = delta + Math.abs(client - server)
                    break
                }
                case "Energy": {
                    this.deltaEnergy = delta + Math.abs(client - server)
                    break
                }
            }
        }
    }
})

var _logState;
LogState.getInstance = function () {
    if (_logState === undefined)
        _logState = new LogState();
    return _logState;
}

File: server/src/logState/LogState.java
package logState;

import battle.GameMgr;
import battle.model.Obstacle;
import battle.model.shootableMonster.Monster;
import battle.model.tower.tower.Tower;
import battle.model.tower.towerAttack.TowerAttack;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.util.ArrayList;

public class LogState {

    private static LogState single_instance = null;

    public static LogState getInstance() {
        if (single_instance == null) {
            single_instance = new LogState();
        }
        return single_instance;
    }

    public void logToFile(GameMgr gameMgr, int countLoop){
        String content = "";

        // Log monster
        content += "Number of monster: " + gameMgr.listMonster.size();
        content += "\n   [id - x - y - current speed - current HP]";
        double sumMonster = 0;
        ArrayList<Monster> arrMonster = gameMgr.getListMonsterInNeighborCell(2, 5, 10);
        for (int i = 0; i < arrMonster.size(); i++) {
            content += "\n" + i + ". ["
                    + arrMonster.get(i).id + " - "
                    + String.format("%.3f", arrMonster.get(i).position.getX()) + " - "
                    + String.format("%.3f", arrMonster.get(i).position.getY()) + " - "
                    + String.format("%.3f", arrMonster.get(i).curSpeed) + " - "
                    + String.format("%.3f", arrMonster.get(i).currentHP)+ "]";

            sumMonster += arrMonster.get(i).id
                    + arrMonster.get(i).position.getX()
                    + arrMonster.get(i).position.getY()
                    + arrMonster.get(i).curSpeed
                    + arrMonster.get(i).currentHP;
        }

        // Log obstacle
        content += "\n\n\nNumber of tree: " + gameMgr.listTreeObstacle.size();
        content += "\n   [id - x - y - current HP]";
        double sumTree = 0;
        ArrayList<Obstacle> arrTree = gameMgr.listTreeObstacle;
        for (int i = 0; i < arrTree.size(); i++) {
            content += "\n" + i + ". ["
                    + arrTree.get(i).id + " - "
                    + String.format("%.3f", arrTree.get(i).position.getX()) + " - "
                    + String.format("%.3f", arrTree.get(i).position.getY()) + " - "
                    + String.format("%.3f", arrTree.get(i).currentHP)+ "]";

            sumTree += arrTree.get(i).id
                    + arrTree.get(i).position.getX()
                    + arrTree.get(i).position.getY()
                    + arrTree.get(i).currentHP;
        }

        // Log tower
        content += "\n\n\nNumber of tower: " + gameMgr.listTower.size();
        content += "\n   [towerID - x - y - evolution]";
        double sumTower = 0;
        for (int i = 0; i < gameMgr.listTower.size(); i++) {
            content += "\n" + i + ". ["
                    + gameMgr.listTower.get(i).towerID + " - "
                    + String.format("%.3f", gameMgr.listTower.get(i).x) + " - "
                    + String.format("%.3f", gameMgr.listTower.get(i).y) + " - "
                    + gameMgr.listTower.get(i).evolution+ "]";

            sumTower += gameMgr.listTower.get(i).towerID
                    + gameMgr.listTower.get(i).x
                    + gameMgr.listTower.get(i).y
                    + gameMgr.listTower.get(i).evolution;

            Tower tower = gameMgr.listTower.get(i);
            if (tower instanceof TowerAttack) {
                int targetID = (((TowerAttack) tower).getTarget() instanceof Monster ? ((Monster) ((TowerAttack) tower).getTarget()).getId() : -1);
                content += " - target: " + targetID;
                sumTower += targetID;
                content += " - range: " + String.format("%.3f", tower.getRange());
                sumTower += tower.getRange();
                content += " - damage: " + String.format("%.3f", ((TowerAttack) tower).getDamage());
                sumTower += ((TowerAttack) tower).getDamage();
                content += " - attackSpeed: " + String.format("%.3f", ((TowerAttack) tower).getAttackSpeed());
                sumTower += ((TowerAttack) tower).getAttackSpeed();
            }
        }

        // Log bullet
        content += "\n\n\nNumber of bullet: " + gameMgr.listBullet.size();
        content += "\n   [x - y - damage - speed]";
        double sumBullet = 0;
        for (int i = 0; i < gameMgr.listBullet.size(); i++) {
            content += "\n" + i + ". ["
                    + String.format("%.3f",gameMgr.listBullet.get(i).x) + " - "
                    + String.format("%.3f",gameMgr.listBullet.get(i).y) + " - "
                    + String.format("%.3f",gameMgr.listBullet.get(i).damage) + " - "
                    + String.format("%.3f",gameMgr.listBullet.get(i).speed) + "]";

            sumBullet += gameMgr.listBullet.get(i).x
                    + gameMgr.listBullet.get(i).y
                    + gameMgr.listBullet.get(i).damage
                    + gameMgr.listBullet.get(i).speed;
        }

        // Round
        String contentSum = String.format("%.3f",sumMonster) + "\n"
                            + String.format("%.3f",sumTree) + "\n"
                            + String.format("%.3f",sumTower) + "\n"
                            + String.format("%.3f",sumBullet) + "\n"
                            + gameMgr.currentEnergy;

        logState(content, gameMgr.user.getId(), countLoop);
        logSumState(contentSum, gameMgr.user.getId(), countLoop);
    }

    public void logState(String content, int id, int countLoop) {
        try {
            // Make new directory
            String dir = "../logServer/logState/user" + id;
            File theDir = new File(dir);
            if (!theDir.exists()){
                theDir.mkdirs();
            }

            // Make new file
            String path = dir + "/"+ countLoop + "-server-stateFrame.txt";
            File file = new File(path);
            file.createNewFile();

            // Write file
            FileWriter fw = new FileWriter(path, false);// ghi de
            BufferedWriter bw = new BufferedWriter(fw);
            bw.write(content);
            bw.close();

        } catch (Exception e) {
            System.out.println(e);
        }
    }

    public void logSumState(String content, int id, int countLoop) {
        try {
            // Make new directory
            String dir = "../logServer/logSumState/user" + id;
            File theDir = new File(dir);
            if (!theDir.exists()){
                theDir.mkdirs();
            }

            // Make new file
            String path = dir + "/" + countLoop + "-server-sumStateFrame.txt";
            File file = new File(path);
            file.createNewFile();

            // Write file
            FileWriter fw = new FileWriter(path, false);// ghi de
            BufferedWriter bw = new BufferedWriter(fw);
            bw.write(content);
            bw.close();

        } catch (Exception e) {
            System.out.println(e);
        }
    }
}
